### ISSUE TITLE
AddressSanitizer: heap-use-after-free mruby/src/vm.c:2137:7

### ISSUE THREAD

**Issue by revskills** [mruby/mruby#3241]
Repro file
base64: bW9kdWxlIFJzCmNsYXNzIFJzY2xhc3MKZGVmICRhLm0KZkBhLmZAYS5tICokQ2QKICAgIGVuZApm
JGEubQogZW5kCmVuZAo=

stacktrace
```sh
$mruby --version
mruby 1.2.0 (2015-11-17) # seems to be the stable version, right?

==31533==ERROR: AddressSanitizer: heap-use-after-free on address 0x61d000001250 at pc 0x0000005b26d3 bp 0x7fffe8c5a090 sp 0x7fffe8c5a088
READ of size 8 at 0x61d000001250 thread T0
    #0 0x5b26d2 in mrb_vm_exec /home/fuzzer/git/mruby/src/vm.c:2137:7
    #1 0x592b6a in mrb_vm_run /home/fuzzer/git/mruby/src/vm.c:766:10
    #2 0x5bd377 in mrb_top_run /home/fuzzer/git/mruby/src/vm.c:2452:12
    #3 0x64bde8 in load_exec /home/fuzzer/git/mruby/mrbgems/mruby-compiler/core/parse.y:5711:7
    #4 0x64a775 in mrb_load_file_cxt /home/fuzzer/git/mruby/mrbgems/mruby-compiler/core/parse.y:5720:10
    #5 0x4f28c7 in main /home/fuzzer/git/mruby/mrbgems/mruby-bin-mruby/tools/mruby/mruby.c:226:9
    #6 0x7fa3eab7e82f in __libc_start_main (/lib/x86_64-linux-gnu/libc.so.6+0x2082f)
    #7 0x41a0e8 in _start (/home/fuzzer/git/mruby/build/test/bin/mruby+0x41a0e8)

0x61d000001250 is located 2000 bytes inside of 2048-byte region [0x61d000000a80,0x61d000001280)
freed by thread T0 here:
    #0 0x4c4ffe in realloc /home/fuzzer/llvm/projects/compiler-rt/lib/asan/asan_malloc_linux.cc:77:3
    #1 0x559675 in mrb_default_allocf /home/fuzzer/git/mruby/src/state.c:60:12
    #2 0x4f4046 in mrb_realloc_simple /home/fuzzer/git/mruby/src/gc.c:201:8
    #3 0x4f4693 in mrb_realloc /home/fuzzer/git/mruby/src/gc.c:215:8
    #4 0x5bdb22 in stack_extend_alloc /home/fuzzer/git/mruby/src/vm.c:151:33
    #5 0x58c831 in stack_extend /home/fuzzer/git/mruby/src/vm.c:168:5
    #6 0x58a40d in mrb_funcall_with_block /home/fuzzer/git/mruby/src/vm.c:382:7
    #7 0x588b7d in mrb_funcall_argv /home/fuzzer/git/mruby/src/vm.c:424:10
    #8 0x58869f in mrb_funcall /home/fuzzer/git/mruby/src/vm.c:319:10
    #9 0x60ba4b in mrb_ary_splat /home/fuzzer/git/mruby/src/array.c:904:7
    #10 0x5b25e3 in mrb_vm_exec /home/fuzzer/git/mruby/src/vm.c:2138:22
    #11 0x592b6a in mrb_vm_run /home/fuzzer/git/mruby/src/vm.c:766:10
    #12 0x5bd377 in mrb_top_run /home/fuzzer/git/mruby/src/vm.c:2452:12
    #13 0x64bde8 in load_exec /home/fuzzer/git/mruby/mrbgems/mruby-compiler/core/parse.y:5711:7
    #14 0x64a775 in mrb_load_file_cxt /home/fuzzer/git/mruby/mrbgems/mruby-compiler/core/parse.y:5720:10
    #15 0x4f28c7 in main /home/fuzzer/git/mruby/mrbgems/mruby-bin-mruby/tools/mruby/mruby.c:226:9
    #16 0x7fa3eab7e82f in __libc_start_main (/lib/x86_64-linux-gnu/libc.so.6+0x2082f)

previously allocated by thread T0 here:
    #0 0x4c4ffe in realloc /home/fuzzer/llvm/projects/compiler-rt/lib/asan/asan_malloc_linux.cc:77:3
    #1 0x559675 in mrb_default_allocf /home/fuzzer/git/mruby/src/state.c:60:12
    #2 0x4f4046 in mrb_realloc_simple /home/fuzzer/git/mruby/src/gc.c:201:8
    #3 0x4f4693 in mrb_realloc /home/fuzzer/git/mruby/src/gc.c:215:8
    #4 0x4f4e53 in mrb_malloc /home/fuzzer/git/mruby/src/gc.c:236:10
    #5 0x4f4eed in mrb_calloc /home/fuzzer/git/mruby/src/gc.c:254:9
    #6 0x58b699 in stack_init /home/fuzzer/git/mruby/src/vm.c:88:28
    #7 0x589886 in mrb_funcall_with_block /home/fuzzer/git/mruby/src/vm.c:356:7
    #8 0x5892a0 in mrb_funcall_with_block /home/fuzzer/git/mruby/src/vm.c:334:13
    #9 0x588b7d in mrb_funcall_argv /home/fuzzer/git/mruby/src/vm.c:424:10
    #10 0x5d3d58 in mrb_obj_new /home/fuzzer/git/mruby/src/class.c:1334:3
    #11 0x66694f in mrb_exc_new_str /home/fuzzer/git/mruby/src/error.c:32:10
    #12 0x66fd3d in mrb_init_exception /home/fuzzer/git/mruby/src/error.c:532:20
    #13 0x675530 in mrb_init_core /home/fuzzer/git/mruby/src/init.c:41:3
    #14 0x559615 in mrb_open_core /home/fuzzer/git/mruby/src/state.c:47:3
    #15 0x5597bc in mrb_open_allocf /home/fuzzer/git/mruby/src/state.c:107:20
    #16 0x55978a in mrb_open /home/fuzzer/git/mruby/src/state.c:99:20
    #17 0x4f1951 in main /home/fuzzer/git/mruby/mrbgems/mruby-bin-mruby/tools/mruby/mruby.c:171:20
    #18 0x7fa3eab7e82f in __libc_start_main (/lib/x86_64-linux-gnu/libc.so.6+0x2082f)

SUMMARY: AddressSanitizer: heap-use-after-free /home/fuzzer/git/mruby/src/vm.c:2137:7 in mrb_vm_exec
Shadow bytes around the buggy address:
  0x0c3a7fff81f0: fd fd fd fd fd fd fd fd fd fd fd fd fd fd fd fd
  0x0c3a7fff8200: fd fd fd fd fd fd fd fd fd fd fd fd fd fd fd fd
  0x0c3a7fff8210: fd fd fd fd fd fd fd fd fd fd fd fd fd fd fd fd
  0x0c3a7fff8220: fd fd fd fd fd fd fd fd fd fd fd fd fd fd fd fd
  0x0c3a7fff8230: fd fd fd fd fd fd fd fd fd fd fd fd fd fd fd fd
=>0x0c3a7fff8240: fd fd fd fd fd fd fd fd fd fd[fd]fd fd fd fd fd
  0x0c3a7fff8250: fa fa fa fa fa fa fa fa fa fa fa fa fa fa fa fa
  0x0c3a7fff8260: fa fa fa fa fa fa fa fa fa fa fa fa fa fa fa fa
  0x0c3a7fff8270: fa fa fa fa fa fa fa fa fa fa fa fa fa fa fa fa
  0x0c3a7fff8280: fa fa fa fa fa fa fa fa fa fa fa fa fa fa fa fa
  0x0c3a7fff8290: fd fd fd fd fd fd fd fd fd fd fd fd fd fd fd fd
Shadow byte legend (one shadow byte represents 8 application bytes):
  Addressable:           00
  Partially addressable: 01 02 03 04 05 06 07
  Heap left redzone:       fa
  Freed heap region:       fd
  Stack left redzone:      f1
  Stack mid redzone:       f2
  Stack right redzone:     f3
  Stack after return:      f5
  Stack use after scope:   f8
  Global redzone:          f9
  Global init order:       f6
  Poisoned by user:        f7
  Container overflow:      fc
  Array cookie:            ac
  Intra object redzone:    bb
  ASan internal:           fe
  Left alloca redzone:     ca
  Right alloca redzone:    cb
==31533==ABORTING
```

**Comment by matz** [comment]
I think the issue has been already fixed by #3151.

Did you try with the latest?  Show us SHA1 hash value of the top commit of your mruby.

**Comment by matz** [comment]
I decode base64 and got the script. I could not reproduce the problem on the latest.
FYI, the program can be reformed to
``` rb
module Rs
  class Rsclass
    def nil.m
      nil.m
    end
    nil.m
  end
end
```
which causes infinite recursion, and the latest detect the stack overflow.
